### ISSUE TITLE
card draw animations become less smooth when battlefield has many cards

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#1127]
I noticed that as the battlefield gets fuller the card draw animations seem to get less smooth. I am guessing it could well be due to ``ImagePermanentViewer.paintComponent()``. You cannot legislate when this is called but even moving the mouse pointer over a card will cause this to fire. Currently it appears to build everything from scratch each time. Instead ``paintComponent`` should do the bare minimum which in this case is draw a cached buffered image that is only re-created each UI update in the game loop.

**Comment by melvinzhang** [comment]
What's the evidence that points to `ImagePermanentViewer.paintComponent()` being the reason for animations getting less smooth?

**Comment by melvinzhang** [comment]
One way to test the hypothesis would be to stub out paintComponent so it doesn't do any work and see if draw animations become smoother. It is important to confirm this method as the culprit before we get into the specifics of how to optimize it.

**Comment by lodici** [comment]
> What's the evidence that points to ImagePermanentViewer.paintComponent() being the reason for animations getting less smooth?

Anecdotal but either way since we cannot know when paintComponent is called it is best practice to ensure that it does the minimum amount of work required. Currently it looks like it is drawing a lot of stuff it does not need to every time unless the game state has changed or the battlefield is resized.

**Comment by melvinzhang** [comment]
True, we cannot know when is called, that why I suggest stubbing it out and seeing what is the effect on the animation. Sometime ago, I did a few rounds of AI speed tuning as faster AI means it can consider more options in the same time. The bottlenecks found by profiling were always unexpected, and did not match my intuition.

**Comment by lodici** [comment]
TestAbilities seems like a good way of stress-testing (see also #1131). What I have found is that it completely maxes out my CPU and is very loathe to release it, as I write this it is waiting for me to declare blockers but CPU is still being hammered. About two or three minutes later CPU has dropped to near zero. 

Probably one of the main drawbacks of Swing is that CPU usage can have significant impact of on the UI. In this case because the CPU is running at 100% _during my turn or my priority_ everything feels sluggish and the card animations are only able to display two or three frames of the animation. Not a very pleasant experience at all.

**Comment by melvinzhang** [comment]
The CPU usage is from the AI trying to compute possible blocks for its attack, Swing drawing only uses one thread unlike the AI which tries to use all CPU cores.

**Comment by lodici** [comment]
In which case the the AI should behave better, IMO. When it is not its turn it should release CPU or at the very least stop trying to grab every cycle. I don't know how realistic TestAbilities is but if this kind of game state crops up often then I think we have a problem because on my 8 Core 2.89 Ghz Intel Core I7 it is unplayable.

**Comment by melvinzhang** [comment]
Absolutely agree with that, it should not be holding onto CPU resources beyond the allocated time. Reword https://github.com/magarena/magarena/issues/1131 to capture the problem. It used to be a lot worse, where the game would be completely stuck and you cannot proceed or even close the game and need to shut it down via task manager.

**Comment by melvinzhang** [comment]
Does this mean the reason for the animation getting less smooth is due the AI due holding on to the CPU or is it due to paintComponent?

The former would only manifest itself after an AI turn, while the latter would always cause the animation to be laggy all the time. @lodici does the less smooth animation follow an AI turn or it happens all the time once there are many cards?

**Comment by lodici** [comment]
Well this test has a large number of permanents on the battlefield over which the draw card animation runs so each frame would require a redraw of the underlying battelfield - whether Swing forces a complete redraw or just the area covered by the animation I am not sure yet. By stubbing out ``battlefieldPanel.doUpdate();`` in ``DuelPanel.update`` you can suppress drawing of the battlefield elements. Currently, because of the AI hogging the CPU it has no affect, animation is poor stubbed or not.

One thing I do notice is that when DA, the permanents are highlighted gradually instead of all at once, likewise when the highlight is removed. This seems to be even when no AI activity. I also get the odd weird painting issues with some cards suddenly being highlighted completely in red or green.

![screenshot00614](https://cloud.githubusercontent.com/assets/2443519/26282936/0c1fc178-3e14-11e7-9000-e144e98003a2.jpg)

**Comment by lodici** [comment]
I changed the AI to MMABFast level 1 so it should have less affect and found there is a distinct improvement in animation when the battlefield drawing is stubbed out. I would be interested in whether @ShawnieBoy or yourself have the same experience.

**Comment by lodici** [comment]
What I would say is that if the above image was a single image on the battlefield instead of 100+ separate ImagePermanentViewers as it is now, I am pretty sure that Swing would be able to optimize by only needing to redraw the area covered, where-as now even if the animation only clips the corner of an ImagePermanentViewer it will still trigger that component's PaintComponent() causing a full redraw of the related card image.

**Comment by lodici** [comment]
[0001-implement-cached-image-in-ImagePermanentViewer.patch.txt](https://github.com/magarena/magarena/files/1016862/0001-implement-cached-image-in-ImagePermanentViewer.patch.txt)

This patch to ImagePermanentViewer lets you see the difference between using a cached image and drawing everything each time paintComponent() is called. 

Please run an amended TestAbilities with MMABFast at level 1.

By default, ``useCachedImage`` is set to false so you get the same behaviour as currently happens. The permanents will be highlighted/unhighlighted in a very noticeable stagger-like way and the draw card animation will reach about 5 fps.

Set ``userCachedImage`` to true. You should notice that the permanents are highlighed/highlighted in one go and the draw card animation should be nice and smooth (especially if you increase it to your monitor's refresh rate using `the `--fps`` command line parameter).

@melvinzhang, @ShawnieBoy please can you give this a try and see if you have the same results. Thanks.

**Comment by melvinzhang** [comment]
The draw card animation looks smoother with cache on. What's the impact on the memory usage? Highlight on mouse over no longer works with cache on.

**Comment by lodici** [comment]
> What's the impact on the memory usage?

There will be some because an image is being created and cached but as more cards are added to the battlefield the images get smaller so overall it will be negligible. It also open to further optimization because quite often there will be duplicate images displayed on the battlefield. For example, the TestAbilities displays 100+ Accorder Paladin images which are all the same so there is potential for some major optimization there. FYI, you can open the About screen at any time to see what if any the impact on memory usage is.

> Highlight on mouse over no longer works with cache on.

Well this was not meant to be the final solution. You were after evidence that the current battlefield drawing code is affecting animation and this quick patch demonstrates that.

Just to illustrate further. When an animation runs the game state should be frozen. The animation might represent the transition from one state to another but while running I would not expect the UI to receive any new game updates. When the draw card animation runs with a single permanent in play the following happens on the ImagePermanentViewer representing the image on the battlefield.
```
>>>>
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
[31455820] paintComponent()
<<<<
```
The chevrons represent the start and end of the animation. That is just one card! It is obvious that the more work that paintComponent has to do in each instance of PermanentViewerInfo the greater the effect on the rest of the UI including animation as demonstrated.

**Comment by melvinzhang** [comment]
> It is obvious that the more work that paintComponent has to do in each instance of ImagePermanentViewer the greater the effect on the rest of the UI including animation as demonstrated.

Not necessarily because there could be something else doing even more work than the paintComponent method. The question is whether paintComponent is the one doing the most work during the animation. We don't have an answer to that, but we do know that it has a noticeable impact on animation smoothness so it is a worthy target. Nice work with the patch.

> Well this was not meant to be the final solution.

Agree. I did not mean that this patch need to address all the issues. Just something I noticed while trying it out, something to take note for the actual solution.

**Comment by melvinzhang** [comment]
I wasn't sure if a profiler would work but I gave visualvm a shot, I sampled the CPU during the draw animation. This is what I got.

![screenshot_2017-05-22_20-47-48](https://cloud.githubusercontent.com/assets/1620634/26309555/0f237a82-3f30-11e7-8801-80e8159c95b6.png)

It shows that ImagePermanentViewer.paintComponent() has the largest self time at 16.9%. The next method is ImageHelper.scale at 9.6%. It conclusively shows that ImagePermanentViewer.paintComponent() caused the bulk of the slow down.

**Comment by lodici** [comment]
Added a slightly amended version of the patch that fixes the mouse over and choice highlighting. Now TestAbilities is still very sluggish (even with MMABFast at level 1) but the draw/play card animations are suddenly really smooth.

**Comment by melvinzhang** [comment]
Looks good, while we are here, another optimization to consider is to only create the image in paintComponent and do not draw the image before calling repaint in redrawCachedImage.

This is because Swing's RepaintManager is able to consolidate multiple repaint calls to reduce the number of actual draws needed, for example of multiple calls to repaint occurs in a short time. However, by drawing before calling repaint, the reduced drawing provided by RepaintManager does not occur. One solution is to mark the viewer as dirty before repaint and check for the dirty flag in the paintComponent method.

**Comment by lodici** [comment]
Good call. See also #1132.

**Comment by lodici** [comment]
Implemented in 1.85.
